src/blockstore/insert_shred.zig

```diff
@@ -0,0 +1,1890 @@
+            std.debug.assert(payload.len == shred_mod.data_shred.payload_size);
+            return payload;
+        }
+    }
+
+    /// agave: has_duplicate_shreds_in_slot
+    fn hasDuplicateShredsInSlot(self: *Self, slot: Slot) !bool {
+        // TODO PERF: just need to check for existence, don't need the value
+        return try self.db.getBytes(schema.duplicate_slots, slot) != null;
+    }
+
+    /// agave: check_merkle_root_consistency
+    fn checkMerkleRootConsistency(
+        self: *Self,
+        just_inserted_shreds: *const AutoHashMap(ShredId, Shred),
+        slot: Slot,
+        merkle_root_meta: *const meta.MerkleRootMeta,
+        shred: *const Shred,
+        duplicate_shreds: *ArrayList(PossibleDuplicateShred),
+    ) !bool {
+        const new_merkle_root = shred.merkleRoot() catch null;
+        if (new_merkle_root == null and merkle_root_meta.merkle_root == null or
+            new_merkle_root != null and merkle_root_meta.merkle_root != null and
+            std.mem.eql(u8, &merkle_root_meta.merkle_root.?.data, &new_merkle_root.?.data))
+        {
+            // No conflict, either both merkle shreds with same merkle root
+            // or both legacy shreds with merkle_root `None`
+            return true;
+        }
+
+        self.logger.warnf(
+            \\Received conflicting merkle roots for slot: {}, erasure_set: {any} original merkle \
+            \\root meta {any} vs conflicting merkle root {any} shred index {} type {any}. Reporting \
+            \\as duplicate
+        , .{
+            slot,
+            shred.commonHeader().erasureSetId(),
+            merkle_root_meta,
+            new_merkle_root,
+            shred.commonHeader().index,
+            shred,
+        });
+
+        if (!try self.hasDuplicateShredsInSlot(slot)) {
+            const shred_id = ShredId{
+                .slot = slot,
+                .index = merkle_root_meta.first_received_shred_index,
+                .shred_type = merkle_root_meta.first_received_shred_type,
+            };
+            if (try self.getShredFromJustInsertedOrDb(just_inserted_shreds, shred_id)) |conflicting_shred| {
+                try duplicate_shreds.append(.{
+                    .MerkleRootConflict = .{
+                        .original = shred.*, // TODO lifetimes (cloned in rust)
+                        .conflict = conflicting_shred,
+                    },
+                });
+            } else {
+                self.logger.errf(
+                    \\Shred {any} indiciated by merkle root meta {any} is 
+                    \\missing from blockstore. This should only happen in extreme cases where 
+                    \\blockstore cleanup has caught up to the root. Skipping the merkle root 
+                    \\consistency check
+                ,
+                    .{ shred_id, merkle_root_meta },
+                );
+                return true;
+            }
+        }
+        return false;
```

Comment:
if you dont want to do it this way, thats fine - but i think adding a comment explaining the control flow would add a lot of value because of the branches + nesting + early returns